### PR TITLE
perf(muse-glimmer): selective 3.5-bit FFN gate/up weights (+5.69% decode)

### PR DESCRIPTION
**Pull request by andriypolanski**
## Summary

Adds a Muse Glimmer-specific, calibration-selected 3.5-bit representation for dense-FFN gate and up projections. Forty-two accuracy-tolerant layers use the compact format while ten sensitive layers remain native Q4_K. The branch includes the decode optimizations merged in #782 and improves 128-token autoregressive decode from 104.36 to 110.66 tok/s (+6.04%) on an RTX 5090 while passing the production llama.cpp distribution gate.

## Proof of speedup

- [x] Tested on **RTX 5090** (`sm_120`)

**Decode tok/s** (end-to-end, three alternating same-binary runs):

| | decode tok/s |
|---|--:|
| before (main after #782; Q3_A disabled) | 104.36 |
| after (this PR; Q3_A enabled) | 110.66 |

```text
# RTX 5090, official 17 GB Muse Glimmer GGUF, 128-token autoregressive decode
# Branch: 235d6f7 (includes #782); only SPARKINFER_MUSE_FFN_Q3A changes between arms.

run      control    this PR
1        104.34     110.61 tok/s
2        104.38     110.70 tok/s
3        104.36     110.68 tok/s
mean     104.36     110.66 tok/s
delta               +6.04%
```

## Design

At model load, selected Q4_K gate/up tensors are dequantized once and fitted into an internal Q3_A super-block. Q3_A keeps Q4_K's asymmetric per-32 scale/min metadata while replacing the 4-bit value plane with a 3-bit plane, reducing each 256-weight block from 144 to 112 bytes.

- Selected layers: 0-26, 33-34, and 39-51.
- Sensitive layers 27-32 and 35-38 remain Q4_K.
- A dedicated dp4a MMVQ kernel consumes equal-stride Q3_A gate/up pairs.
- `SPARKINFER_MUSE_FFN_Q3A=0` restores native Q4_K for same-binary A/B testing.

The format is internal only and is never read from or written to GGUF files. Other architectures retain their existing load and dispatch paths.

## Benchmark details

After merging #782 into this branch, three alternating same-binary runs were performed with the official 17 GB Muse Glimmer GGUF on an RTX 5090. Disabling only Q3_A preserves all #782 optimizations and provides the current-frontier control. The stable means were 104.36 tok/s for control and 110.66 tok/s for selective Q3_A (+6.04%).

```bash
SPARKINFER_MUSE_FFN_Q3A=0 build/runtime/qwen3_gguf_bench model.gguf 128 0
build/runtime/qwen3_gguf_bench model.gguf 128 0
```

## Accuracy

Teacher-forced scoring used the evaluator's 99-position corpus and top-128 logits against a live llama.cpp server using the same GGUF.

| Metric | Result | Required |
|---|---:|---:|
| Top-1 agreement | 0.969697 | >= 0.90 |
| Mean KL | 0.088731 | <= 0.10 |
| SparkInfer PPL | 2.7561 | informational |

## Validation

- Release build for CUDA `sm_120` at combined commit `235d6f7`: passed.
- `ctest --test-dir build --output-on-failure`: 13/13 passed.
- Muse Glimmer accuracy gate: passed.
- `git diff --check`: passed.

## Scope

- Muse Glimmer dense FFN only.
- Gate/up projections only; down projection is unchanged.
- No evaluation, workflow, or protected-path changes.
- No changes to other model architectures.

## Fallback

Set `SPARKINFER_MUSE_FFN_Q3A=0` to disable conversion entirely. `SPARKINFER_MUSE_FFN_Q3A_LAYERS` can override the calibrated layer list for diagnostics.

### CODE REVIEW

**Comment by skyrocket2026** [comment]
## sparkinfer museglimmer auto-eval — `eval-museglimmer:S`

| metric | value |
|---|---|
| **label** | `eval-museglimmer:S` |
| scored at | 128-token decode (ctx=0) + 128-ctx prefill — shared regression floor, best of the two |
| PR decode tok/s | 108.44 |
| main decode tok/s | 102.85 |
| decode speedup vs main | **1.05×** (+5.4%) |
| PR prefill@128 pp tok/s | 115.76 |
| main prefill@128 pp tok/s | 116.23 |
| prefill speedup vs main | **1.00×** (-0.4%) |
| accuracy gate | ✅ top1=0.970 (bar >=0.9) · KL=0.0906 (bar <=0.1) |
| qwen3.6 guard | ✅ no regression (decode+prefill, ctx 0/512/4k/16k/32k) |
| PPL sparkinfer / llama.cpp | 2.6881 / 3.0876 |
| Polaris receipt | `06ff95109172d226…` — TDX (Intel hardware attestation) |
| commit | `235d6f7eb` |

ok

<sub>Scored on the pinned eval box vs same-box `origin/main` — 128-token AR decode (ctx=0) AND 128-ctx prefill throughput, from one model load; either dimension regressing is a hard REJECT, but otherwise the reported label is the **better** of the two tiers — a PR that improves just one, with the other flat, still earns credit for that (no DFlash, no long-context beyond 128) — Muse Glimmer's narrow, deliberately strict eval scope. This is informational, not a judgment on your PR: a `none` label just means no measurable Muse Glimmer speedup was verified on either metric, which is expected and fine if that isn't what your change is about. Correctness gated against a live llama.cpp reference on the same GGUF. Also gated on a Qwen3.6 no-regression guard (decode+prefill, ctx 0/512/4k/16k/32k, same box vs main) — Muse Glimmer PRs can touch code shared with other models. Automated — **not merged**; merge manually after review.</sub>

**Comment by andriypolanski** [comment]
after #782 is merged, my benchmarking result will be changed, I think.
I will update soon, hold on plz.

**Comment by andriypolanski** [comment]
I just updated.

**Comment by andriypolanski** [comment]
Hi,  @skyrocket2026 This branch does not conflict with main.

**Comment by skyrocket2026** [comment]
## sparkinfer museglimmer auto-eval — `eval-museglimmer:XL`

| metric | value |
|---|---|
| **label** | `eval-museglimmer:XL` |
| scored at | 128-token decode (ctx=0) + 128-ctx prefill — shared regression floor, best of the two |
| PR decode tok/s | 108.48 |
| main decode tok/s | 102.87 |
| decode speedup vs main | **1.05×** (+5.5%) |
| PR prefill@128 pp tok/s | 3395.21 |
| main prefill@128 pp tok/s | 1105.33 |
| prefill speedup vs main | **3.07×** (+207.2%) |
| accuracy gate | ✅ top1=0.970 (bar >=0.9) · KL=0.0906 (bar <=0.1) |
| qwen3.6 guard | ✅ no regression (decode+prefill, ctx 0/512/4k/16k/32k) |
| PPL sparkinfer / llama.cpp | 2.6881 / 3.0876 |
| Polaris receipt | `1b3f58f9646ed754…` — TDX (Intel hardware attestation) |
| commit | `1dc63d32d` |

ok

<sub>Scored on the pinned eval box vs same-box `origin/main` — 128-token AR decode (ctx=0) AND 128-ctx prefill throughput, from one model load; either dimension regressing is a hard REJECT, but otherwise the reported label is the **better** of the two tiers — a PR that improves just one, with the other flat, still earns credit for that (no DFlash, no long-context beyond 128) — Muse Glimmer's narrow, deliberately strict eval scope. This is informational, not a judgment on your PR: a `none` label just means no measurable Muse Glimmer speedup was verified on either metric, which is expected and fine if that isn't what your change is about. Correctness gated against a live llama.cpp reference on the same GGUF. Also gated on a Qwen3.6 no-regression guard (decode+prefill, ctx 0/512/4k/16k/32k, same box vs main) — Muse Glimmer PRs can touch code shared with other models. Automated — **not merged**; merge manually after review.</sub>

**Comment by skyrocket2026** [comment]
Auto-merged as the round's `museglimmer-merge-first` winner — verified same-box 128-token decode speedup over `main`, accuracy-gated vs llama.cpp.

**Comment by widecloud** [comment]
Heads up — I think #785 broke Muse Glimmer's batched prefill on main (`9d68aed`).

It converts ffn gate/up to Q3_A and frees the Q4_K source, but the prefill weight path only knows Q4_K/Q5_K/Q6_K, so it falls through `launch_gguf_dequant`'s `else /* F32 */` arm and reads 4 B per weight out of a buffer holding 0.4375. On a 5090:

```
SPARKINFER_KV_INT8=0 build/runtime/qwen3_gguf_prefill_check <muse.gguf> 128 16
  -> illegal memory access, TOP1 0/16, KL 10.79754
```

`5f84500` is clean, so it's that one commit. Decode is unaffected — the +5.4% is real and reproduces.

The awkward part is that the fault sits inside the timed region, so it reads as a speedup rather than a failure: the eval scored it `prefill@128 3395.21 vs 1105.33 = 3.07x`, XL, and auto-merged on that. Repaired it's 1096.29. And until main is fixed its prefill baseline stays ~3330, so anything with a working prefill lands at ~40% of it and gets auto-closed — #790 included.

I have a fix if it's useful: a Q3_A decoder for the prefill dequant path. Gets prefill_check back to 14/16 / KL 0.0366 at 128 and 16/16 / KL 0.0082 at 512.

**Comment by skyrocket2026** [comment]
## Merge reverted — eval harness bug, not this PR's code

This PR's own diff is legitimate: a real ~5.5% decode gain from load-time FFN
gate/up requantization (Q4_K → Q3_A on 42 calibrated layers), passing the
accuracy gate at top1=0.97 / KL=0.089. It never touches any prefill kernel or
dispatch path.

The bot scored it `eval-museglimmer:XL` (+207%) and picked it as the round's
merge-first winner over #790 because of a bogus `prefill@128` reading (3395
pp tok/s) that this PR's code cannot produce. Root cause: the eval harness's
`bench_sweep_run` call used `reps=1` (a single timing sample, no averaging)
for the prefill@128 point. That was fine while Muse Glimmer's prefill ran the
slow token-loop path (~1.1s — dispatch jitter was noise-floor), but #787
(merged the same round) cut that to ~0.1s, so ordinary system jitter now
swings a single-sample reading by 3x. Reproduced live: three repeated runs of
the identical binary landed anywhere from 1103 to 3395 pp tok/s.

Fixed in the bot (`reps=1` → `reps=5`, median), and the merge (`9d68aed`) has
been reverted (`745948f`) so this doesn't keep XL-tier credit it didn't earn.
The real, honest result here is a solid `S`-tier decode gain — please
re-open or re-submit against current `main` and it'll get scored correctly
on the next round.

No fault on the contributor's side — this was purely an eval-harness bug.

**Comment by andriypolanski** [comment]
Hi, @skyrocket2026 
So should I resubmit this pr? this pr is already merged, isn't this?

**Comment by skyrocket2026** [comment]
## Update: this is more than benchmark noise

Following up on the revert above. Turned on per-rep instrumentation and confirmed something
worth flagging directly: on current \`main\` (this PR reverted), 5/5 repeated prefill@128
measurements land in a tight 1105-1115 pp tok/s band -- rock solid. On this PR's own branch,
every measurement was wildly inflated, and taking *more* samples made it *worse*, not better
(single sample: 3395 pp tok/s; median of 5: 84723 pp tok/s).

Your diff never touches a prefill kernel or dispatch path -- the Q3A requantization runs once
at model load, before any prefill code executes. But something about it is destabilizing every
subsequent prefill measurement on that binary. That's a real bug worth tracking down before
resubmitting, not just an eval-harness artifact (the harness's own reps=1 -> reps=5 fix is
already shipped, but it alone doesn't explain numbers this far off). Possibly worth checking
for a stray CUDA allocation/free left in a bad state after \`dev_quant_q3a\`'s load-time
cudaMalloc/cudaFree churn, or a correctness issue in the compact kernel that produces
plausible-looking (accuracy gate still passed) but numerically abnormal output that finishes
faster for the wrong reasons.

Happy to take a look together if useful once you have a fix -- the decode-side win here is real
and worth landing correctly.
